chips/stm32f4xx/src/trng.rs

```diff
@@ -0,0 +1,154 @@
+//! True random number generator
+
+use crate::rcc;
+use kernel::common::cells::OptionalCell;
+use kernel::common::registers::{register_bitfields, ReadOnly, ReadWrite};
+use kernel::common::StaticRef;
+use kernel::hil;
+use kernel::hil::entropy::Continue;
+use kernel::ClockInterface;
+use kernel::ReturnCode;
+
+const RNG_BASE: StaticRef<RngRegisters> =
+    unsafe { StaticRef::new(0x5006_0800 as *const RngRegisters) };
+
+#[repr(C)]
+pub struct RngRegisters {
+    cr: ReadWrite<u32, Control::Register>,
+    sr: ReadWrite<u32, Status::Register>,
+    data: ReadOnly<u32, Data::Register>,
+}
+
+register_bitfields![u32,
+    Control [
+        /// Clock error detection
+        CED OFFSET(5) NUMBITS(1) [
+            ENABLE = 0,
+            DISABLE = 1
+        ],
+        /// Interrupt enable
+        IE OFFSET(3) NUMBITS(1) [],
+        /// True random number generator enable
+        RNGEN OFFSET(2) NUMBITS(1) []
+    ],
+    Status [
+        /// Seed error interrupt status
+        SEIS OFFSET(6) NUMBITS(1) [],
+        /// Clock error interrupt status
+        CEIS OFFSET(5) NUMBITS(1) [],
+        /// Seed error current status
+        SECS OFFSET(2) NUMBITS(1) [],
+        /// Clock error current status
+        CECS OFFSET(1) NUMBITS(1) [],
+        /// Data ready
+        DRDY OFFSET(0) NUMBITS(1) []
+    ],
+    Data [
+        /// Random data
+        RNDATA OFFSET(0) NUMBITS(32) []
+    ]
+];
+
+pub static mut TRNG: Trng<'static> = Trng::new();
+
+pub struct Trng<'a> {
+    registers: StaticRef<RngRegisters>,
+    clock: RngClock,
+    client: OptionalCell<&'a dyn hil::entropy::Client32>,
+}
+
+impl<'a> Trng<'a> {
+    const fn new() -> Trng<'a> {
+        Trng {
+            registers: RNG_BASE,
+            clock: RngClock(rcc::PeripheralClock::AHB2(rcc::HCLK2::RNG)),
+            client: OptionalCell::empty(),
+        }
+    }
+
+    pub fn is_enabled_clock(&self) -> bool {
+        self.clock.is_enabled()
+    }
+
+    pub fn enable_clock(&self) {
+        self.clock.enable();
+    }
+
+    pub fn disable_clock(&self) {
+        self.clock.disable();
+    }
+
+    pub fn handle_interrupt(&self) {
+        if self.registers.sr.is_set(Status::SEIS) {
+            self.registers.sr.modify(Status::SEIS::CLEAR);
+            // Restart the rng
+            self.registers.cr.modify(Control::RNGEN::CLEAR);
+            self.registers.cr.modify(Control::RNGEN::SET);
```

Comment:
It makes sense that if there is a seed error, we restart the RNG. But shouldn't we inform the client of the error, as the client is probably blocked waiting for a callback that now will never arrive?

Reply:
Restarting the rng process is the suggested recovery procedure in the reference manual. After restart, the peripheral is supposed to continue generating correct random numbers and the corresponding interrupts, so I don't see the client getting stuck waiting.

Reply:
Although, It seems that I have omitted to also read the data register before restarting in order to avoid having the client read invalid data.

Reply:
that makes sense, thanks